clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java

```diff
@@ -3405,6 +3406,52 @@ MAX_REQUEST_SIZE, ACKS_ALL, MAX_RETRIES, new SenderMetricsRegistry(new Metrics(t
         assertEquals(1, transactionManager.sequenceNumber(tp1).intValue());
     }
 
+    @Test
+    public void testMakeIllegalTransitionFatal() {
```

Comment:
Can we also include a test that the other illegal state transitions are not fatal (apologies if this is already here)

Reply:
@jolshan Good idea. I'll add that.

Reply:
Is this the change on line 3436? This is useful, but I also wanted to check that such foreground errors don't poison the state. (ie transactionManager.hasFatalError is false) Apologies if it was somewhere else and I missed it

Reply:
Added requested tests for foreground errors.